tls/README.md

```diff
@@ -0,0 +1,287 @@
+  pkiInfo, err := tlsmetadatainterfaces.ProcessByLocation(rawData)
+  if err != nil {
+    return nil, fmt.Errorf("transforming raw data %v: %w", o.GetName(), err)
+  }
+
+  ownershipJSONBytes, err := json.MarshalIndent(pkiInfo, "", "    ")
+  if err != nil {
+    return nil, fmt.Errorf("failure marshalling %v.json: %w", o.GetName(), err)
+  }
+  markdown, err := generateOwnershipMarkdown(pkiInfo)
+  if err != nil {
+    return nil, fmt.Errorf("failure marshalling %v.md: %w", o.GetName(), err)
+  }
+  violations := generateViolationJSON(pkiInfo)
+  violationJSONBytes, err := json.MarshalIndent(violations, "", "    ")
+  if err != nil {
+    return nil, fmt.Errorf("failure marshalling %v-violations.json: %w", o.GetName(), err)
+  }
+
+  return tlsmetadata.NewRequirementResult(
+    o.GetName(),
+    ownershipJSONBytes,
+    markdown,
+    violationJSONBytes)
+}
+```
+
+This interface calls two internal functions:
+```golang
+func generateViolationJSON(pkiInfo *certgraphapi.PKIRegistryInfo) *certgraphapi.PKIRegistryInfo {
+  ret := &certgraphapi.PKIRegistryInfo{}
+
+  // Iterate over certificate-key pairs (==secrets)
+  for i := range pkiInfo.CertKeyPairs {
+    curr := pkiInfo.CertKeyPairs[i]
+    owner := curr.CertKeyInfo.OwningJiraComponent
+    // Include in the violations list if has no owner set or owner set to Unknown
+    if len(owner) == 0 || owner == tlsmetadata.UnknownOwner {
+      ret.CertKeyPairs = append(ret.CertKeyPairs, curr)
+    }
+  }
+
+  // Iterate over CA bundles (==configmaps)
+  for i := range pkiInfo.CertificateAuthorityBundles {
+    curr := pkiInfo.CertificateAuthorityBundles[i]
+    owner := curr.CABundleInfo.OwningJiraComponent
+    // Include in the violations list if has no owner set or owner set to Unknown
+    if len(owner) == 0 || owner == tlsmetadata.UnknownOwner {
+      ret.CertificateAuthorityBundles = append(ret.CertificateAuthorityBundles, curr)
+    }
+  }
+
+  return ret
+}
+```
+
+Markdown report can also be customized, see [example `generateOwnershipMarkdown` method](https://github.com/openshift/origin/blob/master/pkg/cmd/update-tls-artifacts/generate-owners/tlsmetadata/ownership/requirement.go#L71-L160) for ownership requirement.
+
+## Enforcing requirements in tests
+
+Along with the "collect tls artifacts" test the e2e test ensures that cluster certificates don't add 
+new certificates violating requirements via the "all registered tls artifacts must have no metadata violation regressions" test.
+This test would collect the TLS registry from the cluster and generate new violation files from it. If this new file 
+would add new violations to known violations this test would fail, safeguarding us 
+from PRs that add new certificates without required metadata across the entire platform. Another 
+test would verify that metadata for actual certificates matches metadata for known certificate locations.
+
+Once a baseline set of metadata is determined the tests would move from flaking to properly failing.
```

Comment:
I plan to enforce ASAP. I think we have sufficient data to do so

Reply:
This will break a lot of optional operators (i.e. OpenShift GitOps or SRIOV operator), so we need a way of finding all jobs first

Reply:
> This will break a lot of optional operators (i.e. OpenShift GitOps or SRIOV operator), so we need a way of finding all jobs first

do you have a link to such a job so we can see the result?

Reply:
Hmm, I was wrong - I noticed the test didn't pass on a cluster with pipelines+gitops operators installed, but probably its due to service serving CAs, which are accounted now.
However [Hypershift](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-hypershift-release-4.15-periodics-e2e-aws-ovn-conformance/1734196905069514752) jobs would certainly fail, but its relatively easy to fix.

Just to be safe we should switch flake to fail in a separate PR so that TRT would revert just the small part

Reply:
> Just to be safe we should switch flake to fail in a separate PR so that TRT would revert just the small part

Yes.  I also identified https://github.com/openshift/origin/pull/28489/commits/21c9d0d755834dd0d3635f0520a179ee4a7df2b7 which I'm prepared to sacrifice for now to get the majority enforcing.